[CORE-9329] `storage`: concurrent housekeeping

WillemKauf opened the pull request:
This PR splits the housekeeping work done in `log_manager::housekeeping_loop()` into two separate background fibers.

The two background fibers are `log_manager::gc_loop()`, and `log_manager::housekeeping_loop()`.

`log_manager::gc_loop()` handles the urgent garbage collection triggered by resource management and disk space alerts- like before, a semaphore being waited upon is triggered by one of these events, and we attempt to reclaim space from the logs via the `log_eviction_stm` in `log->gc()`.

There is another important change in behavior here though- garbage collection of many logs in the `log_manager` are scheduled at once and waited upon. The fast path to acquire units from a housekeeping lock in `log_housekeeping_meta` is attempted, and if the units aren't acquired, it is because a compaction operation is currently taking place for that log. To prevent blocking `gc_loop()`, garbage collection for that log is aborted early.

`log_manager::housekeeping_loop()` is responsible for invoking garbage collection and then the compaction of a log through `log->housekeeping()`. This still has to be performed serially, as we only allocate one block of memory per shard for sliding window compaction.

All of this has the effect that, concurrently, we can be performing garbage collection of `N-1` logs WHILE compacting 1 log, instead of performing everything serially in one background fiber.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Makes urgent garbage collection and compaction of logs separate, concurrent processes.

WillemKauf committed:
`storage`: add `watchdog` to `housekeeping_scan()`

To improve observability for long running compaction processes
which may be starving out urgent garbage collection.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62113-->
<!--940125c6-5af3-4397-ab4e-3b20cb210f3d-->
<!--55a18a48-1fcc-4272-86e1-a8ff4dd6f766-->
<!--b9e71d68-36df-4928-8d37-b7276b1946bd-->
<!--title-->
## Retry command for [Build#62113](https://buildkite.com/redpanda/redpanda/builds/62113)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::TSWithAlreadyCompactedTopic.test_initial_upload
tests/rptest/tests/full_disk_test.py::FullDiskReclaimTest.test_full_disk_triggers_gc
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":1}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--11048a94-1559-40ab-b71f-1b2e412bbe75-->
<!--TEST RESULTS-->
<!--822522a2-94b8-4ab2-b9f0-c41c4bf72987-->
<!--TEST RESULTS-->
<!--0995b10c-7bb8-4293-a353-3ceeb95f7a53-->
<!--TEST RESULTS-->
<!--cc474e16-984a-4164-85ff-dc42c0d985a0-->
<!--TEST RESULTS-->
<!--635a07bb-c673-47f3-ab73-7fbe4726286c-->
<!--TEST RESULTS-->
<!--0559a5a0-2cc8-4428-afd8-ddb52000a736-->
<!--TEST RESULTS-->
<!--1a239fb6-c399-448b-b501-ca7815d3a18b-->
<!--TEST RESULTS-->
<!--591a62cb-09ad-4951-ab94-80667a61ed38-->
<!--TEST RESULTS-->
<!--ffa5d0e8-ce70-4df3-9b91-6a6f581c3b08-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62113">build#62113</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cloud_storage_rpfixture.cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62113#01952a68-d8da-4087-8512-1998e4248139 | FAIL | 0/2 |
| cloud_storage_rpfixture.cloud_storage_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62113#01952a68-d8db-43c3-bd1f-96bd9714f70f | FAIL | 0/2 |
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62113#01952a68-d8db-43c3-bd1f-96bd9714f70f | FLAKY | 1/2 |
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_unmount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd51-43aa-a6b1-5e136fe9afe1 | FLAKY | 1/2 |
| rptest.tests.full_disk_test.FullDiskReclaimTest.test_full_disk_triggers_gc | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd50-4e2e-a708-906d162c1889 | FAIL | 0/20 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_adding_nodes_to_cluster.partition_count=20 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd50-4e4e-b2df-0c88a91695d9 | FLAKY | 1/2 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd51-43aa-a6b1-5e136fe9afe1 | FAIL | 0/20 |
| rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd4f-4ecf-a1d3-8abe7b4bb3c2 | FAIL | 0/20 |
| rptest.tests.shadow_indexing_compacted_topic_test.TSWithAlreadyCompactedTopic.test_initial_upload | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd50-4e2e-a708-906d162c1889 | FAIL | 0/20 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/62113#01952ac4-bd50-4e2e-a708-906d162c1889 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62145">build#62145</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62145#019530a1-2cf2-4df4-bf60-565c6991c9a1 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62145#019530b5-0042-4df9-af9d-5cc1cf724d70 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/62145#019530b5-0043-4bf7-aa1f-578aa6c836c9 | FLAKY | 1/2 |
| rptest.transactions.stream_verifier_test.StreamVerifierTest.test_simple_produce_consume_txn_with_add_node | ducktape | https://buildkite.com/redpanda/redpanda/builds/62145#019530b5-0044-467f-a34e-873267346567 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62421">build#62421</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62421#01954839-1730-4d56-b158-74de8cab13cd | FLAKY | 1/2 |
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_chunks | ducktape | https://buildkite.com/redpanda/redpanda/builds/62421#01954895-323c-4a1e-9c7d-50712ff51661 | FLAKY | 1/2 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62421#01954895-323a-4ace-b031-afb425dd2b77 | FLAKY | 1/2 |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move_x_core.replication_factor=3.unclean_abort=True.recovery=restart_recovery.compacted=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/62421#01954895-323a-4ace-b031-afb425dd2b77 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62421#0195487b-6c10-402a-8316-2fdd57f20ffe | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62442">build#62442</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance | ducktape | https://buildkite.com/redpanda/redpanda/builds/62442#01954b2f-1534-4e7b-b53b-3a5266ae0319 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=2.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62442#01954b2f-1535-4ee5-8688-37dad16d8e36 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62442#01954b2f-1536-475c-9909-dfc0ec3287d0 | FLAKY | 1/2 |
| rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_unavoidable_data_loss | ducktape | https://buildkite.com/redpanda/redpanda/builds/62442#01954b2f-1534-4e7b-b53b-3a5266ae0319 | FLAKY | 1/2 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62442#01954ad0-e5a0-4e01-9fd7-1acf91cb3d64 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62547">build#62547</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62547#01955cdb-fa30-4043-8471-6506f8cbe566 | FLAKY | 1/2 |
| rptest.tests.offset_for_leader_epoch_archival_test.OffsetForLeaderEpochArchivalTest.test_querying_archive | ducktape | https://buildkite.com/redpanda/redpanda/builds/62547#01955cdb-fa2f-4d83-86fb-04e7f499750c | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62635">build#62635</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_cloud_metadata_rpfixture.cluster_cloud_metadata_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62635#01956424-0be4-40f5-8040-0ced73184f0f | FLAKY | 1/2 |
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62635#01956424-0be2-403e-8073-68de294bae8b | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd41-4b2a-ba2e-5ede9c078982 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd42-4ce6-816e-15a10bc15a0b | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd40-418e-aa9e-72167a6404bf | FLAKY | 1/8 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3add-45b4-89f1-6a500b740de2 | FLAKY | 1/6 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd40-4c70-ba69-6a1af813c1b7 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd41-4b2a-ba2e-5ede9c078982 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd42-4ce6-816e-15a10bc15a0b | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd40-418e-aa9e-72167a6404bf | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd40-4c70-ba69-6a1af813c1b7 | FLAKY | 1/6 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3ade-4810-8a80-1ac1675df5f1 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3adc-4e9a-8434-3f54ac43b696 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd42-4ce6-816e-15a10bc15a0b | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956468-cd40-4c70-ba69-6a1af813c1b7 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3ade-4810-8a80-1ac1675df5f1 | FLAKY | 1/2 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3adc-4e9a-8434-3f54ac43b696 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62635#01956483-3add-4ffa-90b2-b01e96c31844 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62822">build#62822</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| distributed_kv_stm_tests_rpunit.distributed_kv_stm_tests_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/62822#019576ef-8498-45ee-b672-9388f33a4db6 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62822#0195774b-22d7-4e5f-bb87-19bdde0a1065 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62822#01957737-4f80-442d-a48b-805c8e199f02 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63010">build#63010</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63010#01958cb2-59fc-4462-baea-d2682ea5ed49 | FLAKY | 1/2 |
| rptest.tests.datalake.transactions_test.DatalakeTransactionTests.test_with_transactions.cloud_storage_type=CloudStorageType.S3.compaction=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63010#01958cc6-7080-4fd7-bd95-e5bfa6c0f7d5 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63010#01958cc6-7082-40c3-a6a7-cd88f43bc6c5 | FLAKY | 1/2 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=2 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63010#01958cc6-7080-4fd7-bd95-e5bfa6c0f7d5 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/63010#01958cc6-7080-4fd7-bd95-e5bfa6c0f7d5 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63315">build#63315</a></summary>


| test_id                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63315#0195aacb-f7e7-4a6d-83ab-a8682a47c3b8 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63315#0195aacf-bf9d-4248-bd3a-8a2fed70aae5 | FLAKY         | 1/2      |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63315#0195aacf-bf9d-4248-bd3a-8a2fed70aae5 | FLAKY         | 1/2      |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_normalize.dataset_type=JSON                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63315#0195aacb-f7e5-44ec-8c03-a2db7ae7653b | FLAKY         | 1/4      |
| rptest.tests.schema_registry_test.SchemaRegistryTest.test_normalize.dataset_type=JSON                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63315#0195aacf-bfa0-4fd6-a97f-90eb9d610f95 | FLAKY         | 1/2      |
</details>
<!--/body-->

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflict

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1540,6 +1541,7 @@ ss::future<> disk_log_impl::rewrite_segment_with_offset_map(
 
 ss::future<> disk_log_impl::gc(gc_config cfg) {
     ss::gate::holder holder{_compaction_housekeeping_gate};
+    auto housekeeping_lock_holder = co_await _housekeeping_lock.get_units();
     co_await do_gc(cfg);
```
I suspect maybe we'll want this to be some `try_gc()` that just tries to take the lock and bails out if it can't. Otherwise I think we can still run into the problem of a given log's compaction blocking its gc (and subsequent logs in the loop)

andrwng replied:
>(and subsequent logs in the loop)

Ah i guess this isn't true since GC all gets scheduled together

WillemKauf replied:
>(and subsequent logs in the loop)

Correct, we won't block other GCs from getting scheduled together, however, we haven't yet solved the problem of a long-winded compaction on a particular log blocking the gc of that log, which would lead to blocking the `ss::when_all()` for all GC futures, which would lead to blocking the entire `gc_loop`...

Future enhancements to scheduling (i.e bailing out of compaction) will be a follow up.

andrwng replied:
>which would lead to blocking the entire gc_loop

Yea this at least could be mitigated with `try_gc()`. I guess the nice thing then is that even if the given log is blocked in compaction, space management can still make progress in the next round of the loop

Bailing out of compaction makes sense too -- in my head it feels like a harder change to get right, but definitely a decent solution

andrwng replied:
Actually now that we're scheduling compactions based on dirty ratios, i wonder if we _should_ prefer to bail out of GC instead of bailing out of compaction -- presumably compaction _should_ also reduce space? (i guess it depends on the data set, dirty doesn't necessarily mean compacts well)

WillemKauf replied:
>I guess the nice thing then is that even if the given log is blocked in compaction, space management can still make progress in the next round of the loop

This could result in garbage collection never happening, if compaction for a log manages to be running everytime `try_gc()` is attempted. Maybe we get lucky and eventually make progress, but I have a feeling this isn't the angle we want to take for the solution.

Urgent gc/space management should take priority over compaction, IMO, until we somehow implement clever heuristics which can tell us when it shouldn't.

WillemKauf replied:
>Actually now that we're scheduling compactions based on dirty ratios, i wonder if we should prefer to bail out of GC instead of bailing out of compaction -- presumably compaction should also reduce space? (i guess it depends on the data set, dirty doesn't necessarily mean compacts well)

Right, this is the heuristics issue I'm alluding to above.

Dirty ratio on its own doesn't tell us much about the pure number of bytes we can reclaim during a compaction run. And neither would something like an approximation of the key set cardinality. But together, combined, I think would be a powerful heuristic. 

Ultimately, the heuristic needs to serve as an approximation for the number of bytes compaction will reclaim, so we can compare it against the number of bytes space management predicts urgent GC can collect, and decide to bail out of compaction/allow compaction to finish accordingly.

andrwng replied:
> This could result in garbage collection never happening, if compaction for a log manages to be running everytime try_gc() is attempted.

But don't we still GC at the start of housekeeping? I see where you're coming from, but I'm a bit skeptical it would actually be a bad policy, given:
- the housekeeping loop will GC data also
- compaction runs one at a time, meaning other logs are guaranteed to make GC progress
- the recent compaction scheduling improvements I think means we'll compact generally less frequently, giving plenty of room to schedule a good GC. It's hard to imagine we get unlucky so many times in a row with try_gc

Whereas I worry that aborting compaction will end up starving compactions because the common case with space management is to always be GCing a little bit.

WillemKauf replied:
> But don't we still GC at the start of housekeeping?

Mm yeah I thought about this harder and I do think the `try_gc()` approach is sound.

At the very least, more attempts at `gc()` is better than EVERY gc fiber waiting on one `gc()`.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -399,18 +400,18 @@ ss::future<> log_manager::housekeeping_loop() {
-              _jitter.next_duration(),
-              std::max(_housekeeping_sem.current(), size_t(1)));
+              _housekeeping_jitter.next_duration(),
+              std::max(_gc_sem.current(), size_t(1)));
```
Hmm feels a bit weird to not be using the housekeeping sem? Can you add a comment in the header explaining how we should be thinking about each semaphore, when to use them, what they protect, etc?

WillemKauf replied:
As mentioned in my other reply, this chunk of the commit history should be squashed.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -355,17 +363,27 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
     while (!_gate.is_closed()) {
         try {
-            co_await housekeeping_loop();
+            if constexpr (job_t == housekeeping_job_t::housekeeping) {
+                co_await housekeeping_loop();
+            } else if constexpr (job_t == housekeeping_job_t::gc) {
+                co_await gc_loop();
+            } else {
+                static_assert(
+                  always_false_v<log_manager::housekeeping_job_t>,
+                  "Invalid housekeeping_job_t");
+            }
```
Hm, just a thought in case you hadn't considered it, we could just make this some `run_loop(std::function<ss::future<>>, std::string_view)` and it'd be a bit less overhead to read and extend (no asserts, no templates, further loops if any can just add one function and be done)

WillemKauf replied:
I do tend to prefer ~~template magic :magic_wand:~~ compile time code, but I'm happy to change this to a passed `std::function` and `std::string_view ctx`.

andrwng replied:
Yea I'm not against the general practice, but I tend to prefer to use it when needed:
- on hot paths, or
- when it drastically makes life easier (e.g. rpc handling with different req/resp types)

Otherwise it can feel like jumping through hoops as a writer and reader for negligible gain.

dotnwat replied:
> Otherwise it can feel like jumping through hoops as a writer and reader for negligible gain.

yeh i'm not opposed to the code generation, but it seems like it is only saving maybe 10 LOC but then we also have the added complexity of a new named housekeeping "type"

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -399,21 +422,61 @@ ss::future<> log_manager::housekeeping_loop() {
             co_await _housekeeping_sem.wait(
               _housekeeping_jitter.next_duration(),
-              std::max(_gc_sem.current(), size_t(1)));
+              std::max(_housekeeping_sem.current(), size_t(1)));
```
Probably meant to keep _housekeeping_sem in the previous commit?

WillemKauf replied:
Oops yeah, this particular chunk of the commit history should be squashed.

WillemKauf commented:
Force push to:
* Squash chunks of the commit history for easier parsing
* Add `disk_log_impl::try_gc()` and use that in place of `log->gc()` in `log_manager::gc_loop()`. The effect of this is that the call to wait on all `gc` futures with `ss::when_all()` in `log_manager::gc_loop()` will not be held up by a `gc()` fiber that is, for example, blocked by a long-running compaction fiber for that log.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1190,6 +1190,7 @@ gc_config disk_log_impl::apply_overrides(gc_config defaults) const {
 
 ss::future<> disk_log_impl::housekeeping(housekeeping_config cfg) {
     ss::gate::holder holder{_compaction_housekeeping_gate};
+    auto housekeeping_lock_holder = co_await _housekeeping_lock.get_units();
```
apply_segment_ms?

WillemKauf replied:
Are you asking if we need to obtain the same lock in `apply_segment_ms` to prevent races?

WillemKauf replied:
Good point yeah these were two functions that weren't concurrent before

dotnwat replied:
sorry i meant to expand that comment to myself after i finished the review. but looks like you deduced what i had intended.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -142,12 +142,12 @@ log_manager::log_manager(
   , _kvstore(kvstore)
```
> storage: split housekeeping/gc utilities in log_manager
> To be specific about the future uses of the semaphore and jitter variables,
> rename them to `_housekeeping_jitter`. Also, add a new `_gc_sem`.

it seems like this commit claims to do 3 things, where as it does 2? I don't know what "split utilities" really means, but I only see a rename and the introduction of gc_sem. 

A commit should ideally just do one thing.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -399,21 +422,61 @@ ss::future<> log_manager::housekeeping_loop() {
-                continue;
+        /*
+         * if it appears that the compaction interval config changed while
+         * we were sleeping then reschedule rather than run immediately.
```
is there a way to spread the splitting of the housekeeping loop across more than one commit?

WillemKauf replied:
I'll take a look 🙂

WillemKauf committed:
`storage`: rename `_jitter` to `_housekeeping_jitter`

Future commits will split the housekeeping loop into two separate
loops, so be specific about the use for the existing `_jitter` variable
here.

WillemKauf committed:
`storage`: add `_gc_sem`

This will be the semaphore used in future commits for signaling
urgent garbage collection is required, and awaited upon in the
garbage collection fiber's loop.

WillemKauf committed:
`storage`: add `log_manager::run_housekeeping_job`

This function will be used in future commits for running
both the `housekeeping_loop()` and `gc_loop()` functions in the
background.

For now, just alter the code to run the existing `housekeeping_loop()`
using this utility.

WillemKauf committed:
`storage`: break `housekeeping_loop()` into two loops

Using `run_housekeeping_job()`, we now kick off two background
fibres to handle urgent garbage collection and housekeeping.

The underlying functions called are `log->gc()`, `log->housekeeping()`,
and `log->apply_segment_ms()`, which have their concurrency
managed by `log_housekeeping_meta::housekeeping_lock`.

WillemKauf commented:
Force push to:
* Split up commit history further
* Obtain `_housekeeping_lock` in `disk_log_impl::apply_segment_ms()` to prevent concurrency with `disk_log_impl::do_gc()`, which were previously not concurrent operations.
* Remove `job_housekeeping_t` and templated code in `log_manager`, add function signature `log_manager::run_housekeeping_job(std::function<ss::future<>()> loop_func, std::string_view ctx)`

WillemKauf commented:
CI flakes don't seem worrisome in context of these changes.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -187,6 +187,7 @@ ss::future<> log_manager::start() {
 ss::future<> log_manager::stop() {
     _abort_source.request_abort();
     _housekeeping_sem.broken();
+    _gc_sem.broken();
```
nit: maybe it's just me, but I find commits that add a simple unused variable (different story for major plumbing, or config properties) not very reviewer/reader friendly. IMO there isn't much cognitive overhead as a reviewer if it's lumped it with the commit that uses it. As is, it's not really a complete idea, so it's difficult to gauge whether its introduction is correct/justified, especially in going through commit history

WillemKauf replied:
I would agree- I was trying to find ways to further split up the commit history per Noah's request above, perhaps this wasn't the best way to do it :sweat_smile:

dotnwat replied:
> I was trying to find ways to further split up the commit history per Noah's request above

Thanks for the effort. I do agree with Andrew, but it's also a judgement call. My best advice is to keep splitting things up as much as possible, and eventually an intuition will form for what works well.

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -2028,6 +2029,11 @@ ss::future<> disk_log_impl::apply_segment_ms() {
     // naturally).
     auto lock = co_await _segments_rolling_lock.get_units();
 
+    // We also need to hold the housekeeping lock, since the garbage collection
+    // and housekeeping loops were broken into two concurrent fibers. This is to
+    // prevent concurrency with gc() specifically.
+    auto housekeeping_lock_holder = co_await _housekeeping_lock.get_units();
```
Curious about the rationale for putting this below the segment roll lock? How should future authors reason about lock ordering to avoid deadlocks? (worth adding a comment)

WillemKauf replied:
`_segments_rolling_lock` is held in the following functions:
* `disk_log_impl::remove()`
* `disk_log_impl::close()`
* `disk_log_impl::do_truncate()`

In `do_truncate()`, two locks need to be held, `_segments_rolling_lock` and `_segment_rewrite_lock`. There, `_segment_rolling_lock` is held first, so I applied the same ordering here.

I suppose it's worth mentioning that we don't call `_housekeeping_lock.broken()` in `close()` or `remove()` like we do with `_segments_rolling_lock`, maybe we should do that after calling `_compaction_housekeeping_gate.close();`...

andrwng commented on src/v/storage/disk_log_impl.h:
```diff
@@ -85,9 +85,16 @@ class disk_log_impl final : public log {
+    // `gc()` related functions.
     ss::future<> housekeeping(housekeeping_config) final;
-    ss::future<> apply_segment_ms() final;
+    // Attempts to hold `_housekeeping_lock`. Cannot be concurrent with
+    // `housekeeping()` or `apply_segment_ms()`.
```
nit: IMO this can easily be mistaken as a warning "Don't call these methods concurrently", which isn't the guidance I think you're going for. I believe the real guidance is that the three methods are thread-safe with respect to one another and may be called concurrently.

dotnwat replied:
>nit: IMO this can easily be mistaken as a warning "Don't call these methods concurrently", which isn't the guidance I think you're going for. I believe the real guidance is that the three methods are thread-safe with respect to one another and may be called concurrently.

alternatively, since we are changing the way log methods are invoked in log_manager.cc, we could control the concurrency there and leave disk_log_impl.cc untouched.

WillemKauf replied:
Move concurrency control to `log_manager`.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -408,21 +417,61 @@ ss::future<> log_manager::housekeeping_loop() {
+        const auto prev_jitter_base = _housekeeping_jitter.base_duration();
         try {
-            const auto prev_jitter_base = _housekeeping_jitter.base_duration();
             co_await _housekeeping_sem.wait(
```
Just noting that the housekeeping sem isn't really being used to signal anything anymore, and may as well just be a sleep_abortable()

WillemKauf replied:
It's being used to signal run-time changes to `log_compaction_interval_ms`, [see here](https://github.com/redpanda-data/redpanda/blob/a02704a91b594d7809dd8528cd4fa7e0ab0eecf4/src/v/storage/log_manager.cc#L149-L153).

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -362,10 +365,11 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
     while (!_gate.is_closed()) {
         try {
-            co_await housekeeping_loop();
+            co_await ss::futurize_invoke(loop_func);
```
if `loop_func` already returns a future, why does it need to be wrapped in futurize_invoke?

WillemKauf replied:
Removed `futurize_invoke()`

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -489,44 +539,12 @@ ss::future<> log_manager::housekeeping_loop() {
                 }
-                co_await log->gc(
-                  gc_config(lowest_ts_to_retain(), _config.retention_bytes()));
+                gc_futs.push_back(log->try_gc(
+                  gc_config(lowest_ts_to_retain(), _config.retention_bytes())));
```
limit how much is done in parallel? could also parallelize this as a follow up. generally speaking it seems like space management can keep up just fine right?

WillemKauf replied:
Limited parallelism to `20` for garbage collection futures here.

WillemKauf commented:
Push to:
* Remove the `_housekeeping_lock` from `disk_log_impl` and push it into the `log_housekeeping_meta` in `log_manager`
* `log_manager::get()` now returns a log and a associated housekeeping lock for use in the `gc_loop()`.
* The `try_gc()` function that was added to `log` and its derived classes is also removed, since it is no longer needed.
* Limit the concurrency of garbage collection in `gc_loop()` using `ss::max_concurrent_for_each()`.

WillemKauf commented:
Didn't realize how far reaching the `log_manager::get()` changes were, my LSP only showed local use in `log_manager` and some minor tests in `storage/`

Could introduce a new function `log_manager::get_log_and_lock()` to avoid the far reaching test changes with `.first`.

WillemKauf commented:
Apologies for the force pushes reviewers: It will probably be easier to just view the latest commit in isolation when reviewing. Thanks!

WillemKauf commented on src/v/storage/log_manager.cc:
```diff
@@ -543,16 +557,34 @@ ss::future<> log_manager::gc_loop() {
-            }
-
-            co_await ss::when_all(gc_futs.begin(), gc_futs.end());
+            static constexpr size_t max_concurrent_gc = 20;
```
this could also easily be a tunable property.

WillemKauf replied:
Then again, it probably doesn't need to be, given that the amount of work done in `gc()` is practically 0.

Eventually making this not a asynchronous function would be nice (the only bit of asynchronous work within `gc()` is in `segment::get_file_timestamp()` due to adjusting bogus timestamps)

dotnwat commented:
@WillemKauf thanks will review soon. also i see there is a merge conflict which would be useful to resolve

WillemKauf commented:
Yup plan to resolve soon. Need https://github.com/redpanda-data/redpanda/pull/25271 merged first as well.

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev` to fix merge conflicts

andrwng commented on src/v/storage/disk_log_impl.h:
```diff
@@ -86,8 +86,8 @@ class disk_log_impl final : public log {
     ss::future<> housekeeping(housekeeping_config) final;
-    ss::future<> apply_segment_ms() final;
     ss::future<> gc(gc_config) final;
+    ss::future<> apply_segment_ms() final;
```
nit: no more changes do log are needed, yea? Would appreciate just getting rid of the log-focused commits entirely if it's not too much trouble

WillemKauf replied:
Rewrote commit history.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -354,17 +390,19 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
         _probe->housekeeping_log_processed();
 
-        // bail out of compaction early in order to get back to gc
+        // bail out of compaction early in order to allow gc fibre to make
+        // progress
```
Do we still need to do this? Compaction doesn't block the GC fiber anymore, does it?

WillemKauf replied:
This return here is probably not super important anymore. We will need to add better bailing mechanisms into the compaction process while it is underway for a log.

andrwng replied:
Is it better for us not to have it? If we're constantly being asked to bail out could that end up stalling compaction?

WillemKauf replied:
Removed.

andrwng reviewed (dismissed):
High level this lgtm. I'd like us to run a few runs of ducktape and maybe a cdt run to feel more comfortable that this doesn't destabilize things

WillemKauf committed:
`log_manager`: add `get_log_meta()`

We will need to access the `log_housekeeping_meta` in a future commit to
have access to both a `housekeeping_lock` and the underlying `log` when
performing housekeeping on a log.

WillemKauf committed:
`storage`: add `housekeeping_lock` to `log_housekeeping_meta`

Concurrency between garbage collection/housekeeping performed on a `log`
will be managed using this lock within functions in `log_manager`.

Lazin commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1551,6 +1551,18 @@ ss::future<> disk_log_impl::rewrite_segment_with_offset_map(
     vlog(gclog.debug, "[{}] Final compacted segment {}", config().ntp(), seg);
 }
 
+ss::future<> disk_log_impl::try_gc(gc_config cfg) {
```
this is not invoked anywhere, correct?

WillemKauf replied:
Sorry, need to rewrite some of the history here after discussing changes made in the last commit with others.

WillemKauf committed:
`storage`: use `gclog` in `run_housekeeping_job()`

This is a more appropriate logger for this context.

WillemKauf committed:
`storage`: add `log_manager_concurrent_housekeeping_and_removal`

WillemKauf commented:
Force push to:
* Rewrite commit history with accepted changes to concurrency management in the `log_manager`
* Add calls to `housekeeping_lock.broken();` throughout shutdown functions in `log_manager`
* Add sanity checks to `log_meta::link.is_linked()` after new scheduling points in housekeeping functions which obtain `housekeeping_lock`
* Add `log_manager_concurrent_housekeeping_and_removal` fixture test

WillemKauf commented:
/cdt

WillemKauf commented:
CDT failures unrelated to PR.

andrwng commented on tests/rptest/tests/full_disk_test.py:
```diff
@@ -342,11 +342,14 @@ def observed_data_size(pred):
             backoff_sec=2)
 
+        # This expectation has changed since breaking the housekeeping loop
+        # and the urgent gc loop into separate fibres. Disk space alerts only
```
nit: context about the previous implementation generally doesn't hold up well in tests. Could remove the first sentence to keep this comment "stateless"

WillemKauf replied:
Removed.

WillemKauf commented on src/v/storage/log_manager.cc:
```diff
@@ -340,6 +371,23 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
         auto ntp_sanitizer_cfg = _config.maybe_get_ntp_sanitizer_config(
           current_log.handle->config().ntp());
+
+        // Obtain housekeeping lock to prevent concurrency of
```
Flagging that above on [L364](https://github.com/redpanda-data/redpanda/blob/31b58f4ad16a6cb99bfab58cd4a24b8639e46753/src/v/storage/log_manager.cc#L364) we are calling into `gc()` without holding lock, thereby not preventing concurrency with `gc()` invoked by the urgent `gc_loop()`. (This was changed in PR https://github.com/redpanda-data/redpanda/pull/25271)

Not sure that this concurrency is an issue, but we could easily prevent it by adding a `try_get_units()` call above L364.

andrwng replied:
Yea this sgtm.

WillemKauf replied:
Added `try_get_units()` call here.

andrwng commented on src/v/storage/log_manager.cc:
```diff
@@ -181,17 +186,33 @@ ss::future<> log_manager::start() {
       _logs, [this](logs_type::value_type& entry) {
+          // Ensure that users of the housekeeping lock don't try to perform
+          // operations on a removed log_housekeeping_meta
+          entry.second->housekeeping_lock.broken();
```
Mentioned this in Slack, it doesn't seem like these are necessary, given they don't actually prevent ongoing housekeeping from happening. At least, before this change it seems like there is already some expectation that ongoing housekeeping can race with this, so I'm inclined to avoid the broken() call unless we think it's actually problematic

WillemKauf replied:
Removed.

andrwng reviewed (commented):
Pretty much lgtm, just a couple tweaks

WillemKauf commented:
Force push to:
* Remove calls to `housekeeping_lock.broken()`
* Amend comment in `full_disk_test.py`
* Add call to `try_get_units()` in `housekeeping_scan()` ahead of calling `log->gc()` to prevent concurrent garbage collection between `housekeeping_loop()` and `gc_loop()`.

dotnwat commented on src/v/storage/log_housekeeping_meta.h:
```diff
@@ -29,6 +30,8 @@ struct log_housekeeping_meta {
     ss::shared_ptr<log> handle;
     bitflags flags{bitflags::none};
     ss::lowres_clock::time_point last_compaction;
+    // Mutex used for handling concurrency between housekeeping and gc fibres
+    mutex housekeeping_lock{"housekeeping_lock"};
```
thanks for moving this here i think it much easier to understand (even though, yes, we really should have designed locking to be in disk_log_impl from the beginning!)

WillemKauf replied:
Agreed.

nvartolomei replied:
Is there some prior context/conversation re. this not being inside `disk_log_impl`? This is where I'd expect this lock to be as the log (disk_log_impl) is in a much better position to decide what operations need mutual exclusion and which not rather than the caller.

WillemKauf replied:
There were a number of offline and online discussions about this. I think ultimately we decided it was simpler to reason about/a nicer implementation with the concurrency control in `log_manager`.

[Noah's recommendation to control concurrency in `log_manager`](https://github.com/redpanda-data/redpanda/pull/25134#discussion_r1979953591) 

[Andrew's comment about how another lock in `disk_log_impl` means authors have more to reason about when making future changes](https://github.com/redpanda-data/redpanda/pull/25134#discussion_r1978158318)

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -398,21 +448,61 @@ ss::future<> log_manager::housekeeping_loop() {
+     */
+    while (true) {
+        try {
+            co_await _gc_sem.wait(std::max(_gc_sem.current(), size_t(1)));
         } catch (const ss::semaphore_timed_out&) {
```
you're catching `timed_out` here, but you aren't waiting with a specified timeout?

WillemKauf replied:
Oops, yeah this is an artifact of copy pasted code from the old impl.

We don't need to be catching anything here, `_gc_sem` will only be triggered by a call to `signal()` in `handle_disk_notification()` or in `trigger_gc()`.

dotnwat commented on src/v/storage/log_manager.cc:
```diff
@@ -329,9 +356,14 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
-            // basis.
-            co_await current_log.handle->gc(
-              gc_config(collection_threshold, _config.retention_bytes()));
+            // basis. Use `try_get_units()` to avoid concurrent garbage
+            // collection with `gc_loop()`.
```
>to avoid concurrent garbage collection with `gc_loop()`.

doesn't quite answer the question why we want to avoid locking here. avoiding concurrent operation is one thing that is clear from the use of the try lock, but why aren't we _waiting_ to acquire the lock? that's the question to answer here.

WillemKauf replied:
Sorry, I can expand the comment in a follow up PR.

If we `try_get_units()` and don't end up with a lock, the only explanation is that `gc_loop()` is currently performing garbage collection on the same log, and there would be no point to calling into `gc()` in the `housekeeping_loop()` here again.

dotnwat replied:
makes sense thanks

dotnwat commented:
yeh lgtm. just a couple comments/questions above. maybe nothing needs to be done
